metadata-ingestion/src/datahub/ingestion/source/feast.py

```diff
@@ -0,0 +1,266 @@
+                image_directory = os.path.join(dirname, "feast_image/")
+
+                image, _ = docker_client.images.build(path=image_directory)
+
+                feast_image = image.id
+
+            docker_client.containers.run(
+                feast_image,
+                f"python3 ingest.py --core_url={quote(self.config.core_url)} --output_path=/out.json",
+                # allow the image to access the core URL if on host
+                network_mode="host",
+                # mount the tempfile so the Docker image has access
+                volumes={
+                    tf.name: {"bind": "/out.json", "mode": "rw"},
+                },
+            )
+
+            ingest = json.load(tf)
+
+            platform = "feast"
+
+            # ingest tables
+            for table in ingest:
+
+                # ingest entities in table
+                for entity in table["entities"]:
+
+                    # create snapshot instance for the entity
+                    entity_snapshot = MLPrimaryKeySnapshot(
+                        urn=builder.make_ml_primary_key_urn(
+                            table["name"], entity["name"]
+                        ),
+                        aspects=[],
+                    )
+
+                    entity_sources = []
+
+                    if entity["batch_source"] is not None:
+                        entity_sources.append(
+                            builder.make_dataset_urn(
+                                entity["batch_source_platform"],
+                                entity["batch_source_name"],
+                                self.config.env,
+                            )
+                        )
+
+                    if entity["stream_source"] is not None:
+                        entity_sources.append(
+                            builder.make_dataset_urn(
+                                entity["stream_source_platform"],
+                                entity["stream_source_name"],
+                                self.config.env,
+                            )
+                        )
+
+                    # append entity name and type
+                    entity_snapshot.aspects.append(
+                        MLPrimaryKeyPropertiesClass(
+                            description=entity["description"],
+                            dataType=self.get_field_type(
+                                entity["type"], entity["name"]
+                            ),
+                            sources=entity_sources,
+                        )
+                    )
+
+                    # make the MCE and workunit
+                    mce = MetadataChangeEvent(proposedSnapshot=entity_snapshot)
+                    wu = MetadataWorkUnit(id=entity["name"], mce=mce)
+                    self.report.report_workunit(wu)
+                    yield wu
+
+                # ingest features in table
+                for feature in table["features"]:
```

Comment:
may be useful to break this into multiple functions so its easier to read...

maybe getSources, getFeatures, etc